JBPM-9238 Stunner - Allow the introduction of emails in Tasks notifications

hasys opened the pull request:
Hi @romartin, @elguardian, @afalhambra, @LuboTerifaj, @gmunozfe,

there is Stunner part of the introduction of emails in Tasks notifications. It is still missing validation (see below Validation part) but it is ready for End to End tests already. Thank you!

### Related PRs are here
* droolsjbpm-knowledge: https://github.com/kiegroup/droolsjbpm-knowledge/pull/445
* JBPM: https://github.com/kiegroup/jbpm/pull/1684
![image](https://user-images.githubusercontent.com/1477262/87779201-b7716680-c82c-11ea-8a73-d73df53de02c.png)

## How Validation is done
* Split emails string by `,` 
* execute [**OWASP Validation Regex**](https://owasp.org/www-community/OWASP_Validation_Regex_Repository) `^[a-zA-Z0-9_+&*-]+(?:\.[a-zA-Z0-9_+&*-]+)*@(?:[a-zA-Z0-9-]+\.)+[a-zA-Z]{2,7}$` on each part.
* If one of the part is not correct according to the validation, show a message on OK button click:
![image](https://user-images.githubusercontent.com/1477262/87779489-2222a200-c82d-11ea-9125-becebc6f29e3.png)

### WAR file for testing
https://drive.google.com/file/d/1xvT7s00RipwKlb8Y25-ZX1rXaKYuIYCM/view?usp=sharing

hasys commented:
Jenkins execute full downstream build

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:
```diff
@@ -89,11 +89,11 @@
         <div class="form-group">
             <div class="col-sm-3" style="margin-left: -20px;">
-                <p>From</p>
+                <p style="margin: 0">From</p>
```
maybe we can fix also the i18n in this PR

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:
```diff
@@ -89,11 +89,11 @@
             </div>
             <div class="col-sm-3">
-                <p>Reply to (optional)
+                <p style="margin: 0">Reply to (optional)
```
maybe we can fix also the i18n in this PR

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:
```diff
@@ -105,15 +105,19 @@
         <div class="form-group">
             <div class="col-sm-3" style="margin-left: -20px;">
-                <p>To: user(s)</p>
+                <p style="margin: 10px 0 0 0">To: user(s)</p>
```
maybe we can fix also the i18n in this PR

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:
```diff
@@ -105,15 +105,19 @@
             </div>
             <div class="col-sm-3">
-                <p>To: group(s)</p>
+                <p style="margin: 10px 0 0 0">To: group(s)</p>
```
maybe we can fix also the i18n in this PR

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:
```diff
@@ -105,15 +105,19 @@
         </div>
     </div>
+    <div class="row">
+        <p style="margin: 10px 0 0 0;">To: email(s)</p>
```
maybe we can fix also the i18n in this PR

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:
```diff
@@ -105,15 +105,19 @@
     </div>
+    <div class="row">
+        <p style="margin: 10px 0 0 0;">To: email(s)</p>
+        <input data-field="emails" type="text" class="form-control" max='99' placeholder='Enter emails separated by comma.' style='width:500px;'>
```
maybe we can fix also the i18n in this PR

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.html:
```diff
@@ -105,15 +105,19 @@
+        <input data-field="emails" type="text" class="form-control" max='99' placeholder='Enter emails separated by comma.' style='width:500px;'>
+    </div>
     <div class="row">
         <input data-field="subject" type="text" class="form-control" max='99' placeholder='Enter the subject here.' style=' width:500px;    margin-top: 20px;'>
```
maybe we can fix also the i18n in this PR

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.java:
```diff
@@ -426,10 +437,10 @@ public void createOrEdit(NotificationWidgetView parent, NotificationRow row) {
 
         if (row.getType() != null) {
-            if (row.getType().equals(NotificationType.NotCompletedNotify)) {
+            if (row.getType().equals(NOT_COMPLETED_NOTIFY)) {
```
by flipping the condition as NOT_COMPLETED_NOTIFY.equals(row.getType()) there is no need for the NPE check above

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/fields/notificationsEditor/widget/NotificationEditorWidgetViewImpl.java:
```diff
@@ -426,10 +437,10 @@ public void createOrEdit(NotificationWidgetView parent, NotificationRow row) {
                 notCompletedInput.checked = true;
                 notStartedInput.checked = false;
-            } else if (row.getType().equals(NotificationType.NotStartedNotify)) {
+            } else if (row.getType().equals(NOT_STARTED_NOTIFY)) {
```
same here, by flipping the condition no need for the NPE check

romartin commented:
@hasys no worries for the builds on this PR yet, as it cannot yet be tested as it depends on some other PR for being merged, right? As commented, let's just wait for the dependant PRs to be merged and then move this to a normal PR, and let's start the review again. 
Does it makes sense? Plz let us know when this happens.
Thanks!!

hasys commented:
Hi @romartin,

yes, I think I finished here and waiting for the next review :) 

What I did after your last review:
* All layout issues I found in Notification Editor are fixed
* All i18n issues in Notification Editor HTML file are fixed, it is all localized now (just missing translations)
* Fixed coverage issues provided by SonarCloud (I did some refactoring on View/Presenter as well since SonarCloud not calculating GwtMocked coverage correctly I made presenter as normal instance and moved logic (I touched) to the presenter, so in View now mostly logic I didn't touch or declarative only methods).

@afalhambra, @LuboTerifaj I didn't provide Full Downstream Build since there is still missing part in Business Central so it is not possible to perform End to End testing until all other PRs will be merged.

Does it make sense? Should we test UI part only when other PRs will be merged? If not, how will we build Business Central? Since Jenkins can't build Full Downstream Build using PRs from different people.

hasys commented:
Jenkins retest please

hasys commented:
Hi @romartin, @LuboTerifaj  I added more tests, and now SonarCloud complains about duplication only, but duplication caused by duplicated Marshallers, so not much I can do here.

See overall status in [previous comment](https://github.com/kiegroup/kie-wb-common/pull/3358#issuecomment-663110955), thank you!

hasys commented:
Hi @romartin, @afalhambra,

all related PRs are merged and I rebased to master and built a WAR for End To End tests, this PR is ready for review. Thank  you! 

For WAR file see at the bottom https://github.com/kiegroup/kie-wb-common/pull/3358#issue-451032496

hasys commented:
Jenkins retest please

hasys commented:
Jenkins execute compile downstream build

hasys commented:
Jenkins execute compile downstream build

romartin commented:
Hey @hasys 

I'd been looking again at code updates, looks great, great job! :+1: 

Just a couple of Qs before finally approving please:

1. Maybe makes sense making the text area used for the body resizable (as an example look at the text area for any name or documentation field)
2. For the email realted fields to users and groups, I understand we're not querying the user management services right? (which is present for BC, and "mocked" to empty value results in Kogito)

Anyway for me this PR is ok to go, but would like to clarify both improvements above to see if makes sense to also include.

Thanks!

hasys commented:
Hi @romartin good questions, thank you!
1. It is a good possible improvements, but since we have a green build I prefer to report it as separate issue since it is not related to this jira issue. If we will have more fixes to do in this PR than we can do it together since change is trivial. But let's concentrate on merge now if no more critical issues found and improve in separate PR to be safe for the release.
2. No, I didn't use any services. If I correct Users and Groups are provided by EAP/Wildfly user/groups but it doesn't provide any emails. So emails field is just a textbox field without any addition services. Let me know if I am wrong and there is a service for emails. Also I think we can add call to the service as separate PR as well.

Thank you for good questions @romartin!

romartin commented:
Jenkins execute full downtream build

romartin commented:
@hasys lemme know when it's ready to merge!!! :+1:  FDB is running

hasys commented:
Hi @romartin, this PR has green Compile Downstream Build so it is ready for merge. Thank you!
